File: src/modules/jira/jql.ts
import { jiraJson, jiraServerUrl } from "./jira";
import { CliError } from "@/server/cli";
import type { JiraIssue, JqlData, JqlConfig } from "./manifest";

export type JiraRawIssue = {
  key: string;
  fields: {
    summary: string;
    status?: { name?: string };
    assignee?: { displayName?: string } | null;
  };
};

export function normalizeIssue(raw: JiraRawIssue, serverUrl: string): JiraIssue {
  const displayName = raw.fields.assignee?.displayName?.trim();
  return {
    key: raw.key,
    summary: raw.fields.summary,
    status: raw.fields.status?.name ?? "Unknown",
    assignee: displayName ? displayName : null,
    url: `${serverUrl}/browse/${raw.key}`,
  };
}

/**
 * Strip a trailing `ORDER BY` clause. jira-cli appends its own ORDER BY, so a trailing one
 * in the user's JQL is a syntax error. The match must ignore `order by` inside quoted string
 * literals (e.g. `summary ~ "sort order by date"`), so we blank quoted spans to a non-whitespace
 * sentinel (\0) — preserving indices, and non-whitespace so the clause's leading `\s+` can't span
 * across a blanked literal — before locating the clause, then cut the original at that index.
 */
export function stripTrailingOrderBy(jql: string): string {
  let masked = "";
  let quote: '"' | "'" | null = null;
  for (let i = 0; i < jql.length; i++) {
    const ch = jql[i];
    if (quote) {
      if (ch === "\\" && i + 1 < jql.length) {
        masked += "\0\0"; // blank the backslash and the char it escapes
        i++;
        continue;
      }
      masked += "\0";
      if (ch === quote) quote = null;
    } else if (ch === '"' || ch === "'") {
      masked += "\0";
      quote = ch;
    } else {
      masked += ch;
    }
  }
  const m = masked.match(/\s+order\s+by\s+[\s\S]+$/i);
  return (m ? jql.slice(0, m.index) : jql).trim();
}

export async function fetchJql(config: JqlConfig): Promise<JqlData> {
  const jql = stripTrailingOrderBy(config.jql);
  try {
    const raw = await jiraJson<JiraRawIssue[]>([
      "issue", "list", "-q", jql, "--order-by", "updated", "--paginate", `0:${config.limit}`,
    ]);
    const server = await jiraServerUrl();
    return { issues: raw.map((r) => normalizeIssue(r, server)) };
  } catch (err) {
    // jira-cli exits non-zero with this message when a query matches nothing.
    if (err instanceof CliError && /no result found/i.test(err.message)) {
      return { issues: [] };
    }
    throw err;
  }
}

File: internal/modules/jira/jql.go
package jira

import (
	"context"
	"errors"
	"fmt"
	"regexp"
	"strings"
	"unicode/utf8"

	"pulse/internal/cli"
)

// Issue mirrors the TS JiraIssue payload shape.
type Issue struct {
	Key      string  `json:"key"`
	Summary  string  `json:"summary"`
	Status   string  `json:"status"`
	Assignee *string `json:"assignee"` // null when unassigned
	URL      string  `json:"url"`      // <server>/browse/<KEY>
}

type JqlData struct {
	Issues []Issue `json:"issues"`
}

type jqlConfig struct {
	Jql   string `json:"jql"`
	Limit int    `json:"limit"`
}

type rawIssue struct {
	Key    string `json:"key"`
	Fields struct {
		Summary string `json:"summary"`
		Status  *struct {
			Name string `json:"name"`
		} `json:"status"`
		Assignee *struct {
			DisplayName string `json:"displayName"`
		} `json:"assignee"`
	} `json:"fields"`
}

var orderByRe = regexp.MustCompile(`(?is)\s+order\s+by\s+.+$`)

// stripTrailingOrderBy strips a trailing ORDER BY clause — jira-cli appends
// its own, so a trailing one in the user's JQL is a syntax error. Quoted
// string literals are blanked to a non-whitespace NUL sentinel (preserving
// rune indices, and non-whitespace so the clause's leading \s+ can't span a
// blanked literal) before locating the clause; the original is cut at the
// found index.
func stripTrailingOrderBy(jql string) string {
	runes := []rune(jql)
	masked := make([]rune, 0, len(runes))
	var quote rune
	for i := 0; i < len(runes); i++ {
		ch := runes[i]
		switch {
		case quote != 0:
			if ch == '\\' && i+1 < len(runes) {
				masked = append(masked, 0, 0) // blank the backslash and the escaped char
				i++
				continue
			}
			masked = append(masked, 0)
			if ch == quote {
				quote = 0
			}
		case ch == '"' || ch == '\'':
			masked = append(masked, 0)
			quote = ch
		default:
			masked = append(masked, ch)
		}
	}
	maskedStr := string(masked)
	loc := orderByRe.FindStringIndex(maskedStr)
	if loc == nil {
		return strings.TrimSpace(jql)
	}
	runeIdx := utf8.RuneCountInString(maskedStr[:loc[0]])
	return strings.TrimSpace(string(runes[:runeIdx]))
}

func normalizeIssue(raw rawIssue, serverURL string) Issue {
	issue := Issue{
		Key: raw.Key, Summary: raw.Fields.Summary, Status: "Unknown",
		URL: serverURL + "/browse/" + raw.Key,
	}
	if raw.Fields.Status != nil && raw.Fields.Status.Name != "" {
		issue.Status = raw.Fields.Status.Name
	}
	if raw.Fields.Assignee != nil {
		if name := strings.TrimSpace(raw.Fields.Assignee.DisplayName); name != "" {
			issue.Assignee = &name
		}
	}
	return issue
}

var noResultRe = regexp.MustCompile(`(?i)no result found`)

func (m *Module) fetchJql(ctx context.Context, cfg jqlConfig) (JqlData, error) {
	jql := stripTrailingOrderBy(cfg.Jql)
	raw, err := jiraJSON[[]rawIssue](ctx, m.run, []string{
		"issue", "list", "-q", jql, "--order-by", "updated", "--paginate", fmt.Sprintf("0:%d", cfg.Limit),
	})
	if err != nil {
		// jira-cli exits non-zero with this message when a query matches nothing.
		var ce *cli.Error
		if errors.As(err, &ce) && noResultRe.MatchString(ce.Message) {
			return JqlData{Issues: []Issue{}}, nil
		}
		return JqlData{}, err
	}
	server, err := m.serverURL()
	if err != nil {
		return JqlData{}, err
	}
	issues := make([]Issue, 0, len(raw))
	for _, r := range raw {
		issues = append(issues, normalizeIssue(r, server))
	}
	return JqlData{Issues: issues}, nil
}
